add fuzzy matching for versions

rmg opened the pull request:
Extend the existing NodeVersion class with a FuzzyNodeVersion class that
overrides the .match() method behaviour to do a fuzzy match against the
given version.

A FuzzyNodeVersion is created when a given "version" input string is
just a simple string that doesn't map to an existing special case (lts,
current, default, latest) and isn't simply a number (does semver
matchign against default remote).

Closes #37

This does _not_ include any sort of documentation since I wanted to get the concept and implementation sorted out first.

jasongin commented:
Thanks for the contribution!

This is a little different from how I would have approached it. I'm concerned that the missing properties on the version object (remoteName, semanticVersion, arch, label) may cause problems in some commands, for example `nvs add`, which looks at those properties instead of just calling `match()`.

It should be possible to fully resolve a partial remoteName to one of the configured remote names at the time the version is parsed, rather than using different matching logic later on. I think there is no need to enable partial matching on the semantic version at parsing time, because the NodeVersion.match()  method already has special semver logic.

Also, can we use the term "partial" instead of "fuzzy"? I think the meaning would be clearer.

rmg commented:
The fact that it doesn't work with `nvs add` is intentional. I attempted to indicate this with the change I made to `NodeVersion.equal` so that a `FuzzyNodeVersion` is never considered equal to a real `NodeVersion`.

It is a little confusing what exactly a `NodeVersion` is meant to encapsulate and where to hang the new functionality and I figured a subclass was the cleanest way to augment it without actually modifying its behaviour too much.

Can you give a little more detail about how you would have done it? I promise I won't be offended if you decide to just do it yourself as a way of illustrating what you wanted :-)

> can we use the term "partial" instead of "fuzzy"? I think the meaning would be clearer.

This is actually intentional! My plan was to eventually implement proper fuzzy matching instead of just partial or substring matching: eg. `n61` matches **n**ode/**6**.**1**.10/x64.

jasongin commented:
> The fact that it doesn't work with nvs add is intentional.

I'm not sure I like having a version syntax that works with some commands but not others.

> My plan was to eventually implement proper fuzzy matching instead of just partial or substring matching: eg. n61 matches node/6.1.10/x64.

Ahh, I didn't realize that. I guess the term "fuzzy" would make a little more sense then. Is that sort of matching inspired by something else? Another version-manager tool?

> Can you give a little more detail about how you would have done it?

Well, I had been thinking of just substring matches. For that I would add a few lines in `NodeVersion.parse()` to map a partial `remoteName` to a complete name from `settings.remotes`. That would enable commands like `nvs use n` or `nvs use n/6.1` (assuming 'n' is not ambiguous). And it would work just as well with `nvs add`, since the resulting `NodeVersion` object is the same regardless of whether it was a partial or full match.

For more "fuzzy" matching, I'm not sure. It depends on how "fuzzy" it needs to be. For a more limited definition, perhaps remote names could be restricted to letters only, so any numbers would imply the start of a semver without dots -- basically the '/' and '.' separators would be implied, then the logic would be the same as the partial matching described above.

rmg commented:
I've been living with this applied to my local `nvs` and haven't felt the urge to tweak it. I'm not sure how much more time I'll put in now that my original itch has more or less been scratched. I'll at least keep this PR updated as `master` progresses, hopefully addressing comments over time.

rmg commented:
I haven't touched this in over a year and I'm not even using these changes locally anymore, so I'm going to close this now.
